### PR TITLE
Update LoadSecrets()

### PR DESCRIPTION
**Pull request by lakshmimsft**
# Description

1. Update LoadSecrets() to return data for all keys for corresponding secret store when no keys filter is provided.
2. Update return type to include SecretData{}
3. Updated existing LoadSecrets() to return data collated for multiple secret store ids
4. Update to helper function populateSecretData() and it's unit tests which populates secret data map returned.
5. Update functional test to include envSecrets input

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (#6917).

Fixes: (Part of #6917)

### CODE REVIEW

File: pkg/recipes/configloader/secrets.go [hunk 1]
```diff
@@ -56,32 +60,49 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
+	for _, secretKey := range secretKeysFilter {
+		secretDataValue, ok := secrets.Data[secretKey]
+		if !ok {
 			return nil, fmt.Errorf("a secret key was not found in secret store '%s'", secretStoreID)
 		}
+
+		if secretData[secretStoreID] == nil {
+			secretData[secretStoreID] = make(map[string]string)
+		}
+		secretData[secretStoreID][secretKey] = *secretDataValue.Value
```
Comment (lakshmimsft) [inline pkg/recipes/configloader/secrets.go:105]: ln96- ln105 is just ln76-ln82 refactored

File: pkg/recipes/configloader/secrets.go [hunk 2]
```diff
@@ -35,10 +35,14 @@ type secretsLoader struct {
...
-// It returns a map of secret data, where the keys are the secret store IDs and the values are maps of secret keys and their corresponding values.
-func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKeys map[string][]string) (secretData map[string]map[string]string, err error) {
-	for secretStoreID, secretKeys := range secretStoreIDResourceKeys {
+// LoadSecrets loads secrets from secret stores based on input map of provided secret store IDs and secret keys filter.
+// If the input secret keys filter for a secret store ID is nil or empty, it retrieves secret data for all keys for that secret store.
+// When the secret keys filter is populated for a secret store ID, it retrieves secret data for the specified keys for the associated secret store ID.
+// The function returns a map of secret data, where the keys are the secret store IDs and the values are maps of secret keys and their corresponding values.
+func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDKeysFilter map[string][]string) (secretData map[string]map[string]string, err error) {
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets.go:42]: Why not just name this `filter` and give an example in the doc above?
```suggestion
func (e *secretsLoader) LoadSecrets(ctx context.Context, filter map[string][]string) (secretData map[string]map[string]string, err error) {
```

File: pkg/recipes/configloader/secrets.go [hunk 3]
```diff
@@ -56,32 +60,49 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
 	return secretData, nil
 }
 
 // populateSecretData is a helper function to populate secret data from a secret store.
-func populateSecretData(secretStoreID string, secretKeys []string, secrets *v20231001preview.SecretStoresClientListSecretsResponse) (map[string]map[string]string, error) {
-	secretData := make(map[string]map[string]string)
-
+// It takes a secret store ID, a filter for secret keys, and a response containing the secrets.
+// If the secret keys filter is nil or empty, it retrieves data for all keys in the secret store.
+// It returns a map where the keys are the secret store IDs and the values are maps of secret keys and their corresponding values.
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets.go:80]: Should start with what the function does and then give examples like you did here.

File: pkg/recipes/configloader/secrets.go [hunk 4]
```diff
@@ -56,32 +60,49 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
+func populateSecretData(secretStoreID string, secretKeysFilter []string, secrets *v20231001preview.SecretStoresClientListSecretsResponse) (map[string]map[string]string, error) {
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets.go:81]:
```suggestion
func populateSecretData(id string, filter []string, secrets *v20231001preview.SecretStoresClientListSecretsResponse) (map[string]map[string]string, error) {
```

File: pkg/recipes/configloader/secrets.go [hunk 5]
```diff
@@ -56,32 +60,49 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
+
+	for _, secretKey := range secretKeysFilter {
+		secretDataValue, ok := secrets.Data[secretKey]
+		if !ok {
 			return nil, fmt.Errorf("a secret key was not found in secret store '%s'", secretStoreID)
 		}
+
+		if secretData[secretStoreID] == nil {
+			secretData[secretStoreID] = make(map[string]string)
+		}
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets.go:104]: This one will be empty for sure, right? Why don't we move this to line 87 right after we initialize the slice? That would be more readable maybe?

File: pkg/recipes/configloader/secrets_test.go [hunk 6]
```diff
@@ -19,8 +19,30 @@ func Test_populateSecretData(t *testing.T) {
 		expectedErrMsg  string
 	}{
 		{
-			name:       "success",
-			secretKeys: []string{"secretKey1", "secretKey2"},
+			name:       "success - data for input secretKey returned",
```
Comment (ytimocin) [inline pkg/recipes/configloader/secrets_test.go:22]: This is actually secretKey1, right?
```suggestion
			name:       "success - data for input secretKey1 returned",
```

File: pkg/recipes/configloader/secrets.go (hunk 2 above)
Comment (vishwahiremat) [inline pkg/recipes/configloader/secrets.go:42]: return type should be updated to add the type of secret.
```
type secretData struct{
  type string
  Data map[string]string
}
```

File: pkg/recipes/configloader/secrets.go [hunk 7]
```diff
@@ -56,32 +60,49 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
+
+	// If secretKeys is nil or empty, retrieve data for all keys
+	if len(secretKeysFilter) == 0 {
+		secretKeysFilter = make([]string, 0, len(secrets.Data))
+		for secretKey := range secrets.Data {
+			secretKeysFilter = append(secretKeysFilter, secretKey)
+		}
+	}
+
+	for _, secretKey := range secretKeysFilter {
```
Comment (vishwahiremat) [inline pkg/recipes/configloader/secrets.go:96]: i think this should be part of `else` block?

Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go:96]: this is correct as is. if condition ln89 populates secretKeysFilter with data for all keys if empty as described.
the for loop ln96 populates the return type for the list of keys.

File: pkg/recipes/configloader/secrets.go (hunk 5 above)
Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go:104]: the return type is a map of maps. we do not want to return a map of keys (secretstoreids)with no data if no secret data exist. Keeping this as is for now.

File: pkg/recipes/configloader/secrets.go (hunk 4 above)
Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go]: keeping the more descriptive names

File: bicep-types
Comment (vishwahiremat) [inline bicep-types:1]: are these changes related to this PR?

File: pkg/recipes/configloader/secrets.go [hunk 8]
```diff
@@ -55,33 +64,55 @@ func (e *secretsLoader) LoadSecrets(ctx context.Context, secretStoreIDResourceKe
...
+		Data: make(map[string]string),
+	}
+
+	// If secretKeysFilter is nil or empty, populate secretKeysFilter with all keys
+	if len(secretKeysFilter) == 0 {
+		secretKeysFilter = make([]string, 0, len(secrets.Data))
+		for secretKey := range secrets.Data {
+			secretKeysFilter = append(secretKeysFilter, secretKey)
+		}
 	}
```
Comment (vishwahiremat) [inline pkg/recipes/configloader/secrets.go:108]: can we do it this way instead of creating a secretKeyFilter map ?
```suggestion
	if len(secretKeysFilter) == 0 {
	        secretData.Data = secrets.Data
	        return secretData,nil
	}
```

Comment (lakshmimsft) [reply pkg/recipes/configloader/secrets.go:108]: secrets.Data is a map[string]*SecretValueProperties, different datatype from what is needed. keeping current code as discussed offline.

File: pkg/recipes/terraform/execute.go [hunk 9]
```diff
@@ -225,7 +225,7 @@ func (e executor) setEnvironmentVariables(tf *tfexec.Terraform, options Options)
 		for secretName, secretReference := range recipeConfig.EnvSecrets {
 			// Extract secret value from the secrets input
 			if secretIDs, ok := options.Secrets[secretReference.Source]; ok {
-				if secretValue, ok := secretIDs[secretReference.Key]; ok {
+				if secretValue, ok := secretIDs.Data[secretReference.Key]; ok {
 					envVarUpdate = true
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply pkg/recipes/terraform/execute.go:229]: secretIDs is a struct and will not be nil. Renamed variable to secretData. There exists a unit test to check for secretData.Data when nil.

File: bicep-types
Comment (lakshmimsft) [reply bicep-types:1]: removed from PR.

File: test/functional-portable/corerp/noncloud/resources/testdata/corerp-resources-terraform-postgres.bicep [hunk 10]
```diff
@@ -24,13 +24,15 @@ resource env 'Applications.Core/environments@2023-10-01-preview' = {
         providers: {
```
[the remark this answers is not in the document]
Comment (lakshmimsft) [reply test/functional-portable/corerp/noncloud/resources/testdata/corerp-resources-terraform-postgres.bicep:24]: Discussed offline. The functional test needs to be updated with envSecrets block. This PR updates code that can affect running of functional test with envSecrets. We'll include this test in the PR.

**Review by vishwahiremat — APPROVED** [review]
lgtm

**Review by ytimocin — APPROVED** [review]
